// parser/lib/parser.js
var fs = require('fs');
var path = require('path');
var csv = require("csv");
var JSONStream = require('JSONStream');
var jsonToStrings = JSONStream.stringify(false);

var parser = module.exports = function parser(options){
    this.type = options.type;
    this.delimiter = options.delimiter;
    this.comment = options.comment;
    this.location = options.location;
};

parser.prototype.getDelimiter = getDelimiter;
parser.prototype.getComment = getComment;
parser.prototype.parse = parse;

function parse(){
    var fstream = fs.createReadStream(this.location);
    var encoding = fstream.defaultEncoding;
    var csvParseOptions =  { objectMode: true , delimiter : this.delimiter, comment : this.comment, columns:true};
    if(!encoding) encoding = "utf8";
    fstream.setEncoding(encoding);
    
    return fstream.pipe(csv.parse(csvParseOptions))
        .pipe(csv.transform(
                //callback per line
                function(record){ 
                    return record;
                }
            ))
        .pipe(jsonToStrings); //{"colA":"valA", "colB":"valB"....};
}

function getDelimiter(){
    return this.delimiter;
};

function getComment(){
    return this.comment;
}
