// parser/lib/parser.js
var fs = require('fs');
var path = require('path');
var csv = require("csv");

var parser = module.exports = function parser(options){
    this.type = options.type;
    this.delimiter = options.delimiter;
    this.comment = options.comment;
    this.location = options.location;
};

parser.prototype.getDelimiter = getDelimiter;
parser.prototype.getComment = getComment;
parser.prototype.parse = parse;

function parse(){
    var fstream = fs.createReadStream(this.location);
    var encoding = fstream.defaultEncoding;
    var csvParseOptions =  { delimiter : this.delimiter, comment : this.comment, columns:true};
    if(!encoding) encoding = "utf8";
    fstream.setEncoding(encoding);
    
    return fstream.pipe(csv.parse(csvParseOptions))
        .pipe(csv.transform(function(record){ //callback per line
                //var line = [];
                //record.map(function(value){ //callback per col
                //        line.push(value);
                //        //line.push({"capv_string":value});
                //    });
                //console.log(line);
                //return line;
                console.log(record)
                return record;
        }))
        .pipe(csv.stringify());
}

function getDelimiter(){
    return this.delimiter;
};

function getComment(){
    return this.comment;
}
